```
error[E0599]: no function or associated item named `unpack` found for type parameter `U` in the current scope
   --> components/datetime/src/provider/packed_pattern.rs:276:12
    |
272 |     fn from_packed<U: VarULE + ?Sized>(packed: &'a GenericPackedPatterns<'a, U>) -> Self
    |                    - function or associated item `unpack` not found for this type parameter
...
276 |         U::unpack(packed)
    |            ^^^^^^ function or associated item not found in `U`
    |
help: there is an associated function `pack` with a similar name
    |
276 -         U::unpack(packed)
276 +         U::pack(packed)

```

--- a/components/datetime/src/provider/packed_pattern.rs
+++ b/components/datetime/src/provider/packed_pattern.rs
@@ -628,6 +628,7 @@ mod _serde {
     trait PackedPatternsSerdeHelper: PackedPatternsBuilderHelper {
         type Human;
         fn human_to_unpacked_element<'a>(human: &'a Self::Human) -> Self::Unpacked<'a>;
+        #[cfg(feature = "datagen")]
         fn unpacked_element_to_human<S: serde::Serializer>(
             element: Self::Unpacked<'_>,
         ) -> Result<Self::Human, S::Error>;
@@ -639,6 +640,7 @@ mod _serde {
             let runtime_pattern = human.to_runtime_pattern();
             PluralElements::new(runtime_pattern)
         }
+        #[cfg(feature = "datagen")]
         fn unpacked_element_to_human<S: serde::Serializer>(
             element: Self::Unpacked<'_>,
         ) -> Result<Self::Human, S::Error> {

--- a/components/datetime/src/provider/packed_pattern.rs
+++ b/components/datetime/src/provider/packed_pattern.rs
@@ -241,9 +241,9 @@ where
 {
     /// Generically builds a `GenericPackedPatterns` structure, using a caller-supplied
     /// closure to pack the elements into a `VarZeroVec`.
-    fn build<U: VarULE + ?Sized>(&self) -> GenericPackedPatterns<'static, U>
+    fn build<U>(&self) -> GenericPackedPatterns<'static, U>
     where
-        U: PackedPatternsBuilderHelper<Unpacked<'a> = T>,
+        U: PackedPatternsBuilderHelper<Unpacked<'a> = T> + VarULE + ?Sized,
     {
         let mut header = 0u32;
         if self.has_explicit_medium {
@@ -269,9 +269,10 @@ where
         GenericPackedPatterns { header, elements }
     }
 
-    fn from_packed<U: VarULE + ?Sized>(packed: &'a GenericPackedPatterns<'a, U>) -> Self
+    #[cfg(feature = "datagen")]
+    fn from_packed<U>(packed: &'a GenericPackedPatterns<'a, U>) -> Self
     where
-        U: PackedPatternsBuilderHelper<Unpacked<'a> = T>,
+        U: PackedPatternsBuilderHelper<Unpacked<'a> = T> + VarULE + ?Sized,
     {
         U::unpack(packed)
     }

